doc/designs/ech-api.md

```diff
@@ -0,0 +1,713 @@
+inner CH. Note that, when ECH decryption succeeds, the inner CH will always be
+smaller than the outer CH.
+
+If there is no ECH present in the outer CH then this will return 1 (i.e., the
+call will succeed) but `decrypted_ok` will be zero. The same will result if a
+GREASEd ECH is present or decryption fails for some other (indistinguishable)
+reason.
+
+If the caller wishes to support HelloRetryRequest (HRR), then it must supply
+the same `hrrtok` and `toklen` pointers to both calls to
+`SSL_CTX_ech_raw_decrypt()` (for the initial and second ClientHello
+messages). When done, the caller must free the `hrrtok` using
+`OPENSSL_free()`.  If the caller doesn't need to support HRR, then it can
+supply NULL values for these parameters. The value of the token is the client's
+ephemeral public value, which is not sensitive having being sent in clear in
+the first ClientHello.  This value is missing from the second ClientHello but
+is needed for ECH decryption.
+
+Note that `SSL_CTX_ech_raw_decrypt()` only takes a ClientHello as input. If
+the flight containing the ClientHello contains other messages (e.g. a
+ChangeCipherSuite or Early data), then the caller is responsible for
+disentangling those, and for assembling a new flight containing the inner
+ClientHello.
+
+Different encodings
+-------------------
+
+ECHConfigList values may be provided via a command line argument to the calling
+application or (more likely) have been retrieved from DNS resource records by
+the application. ECHConfigList values may be provided in various encodings
+(base64, ascii hex or binary) each of which may suit different applications.
+ECHConfigList values may also be provided embedded in the DNS wire encoding of
+HTTPS or SVCB resource records or in the equivalent zone file presentation
+format.
+
+`OSSL_ECHSTORE_find_echconfigs()` attempts to find and return the (possibly empty)
+set of ECHConfigList values as an `OSSL_ECHSTORE` from the input `BIO`.
+
+```c
+OSSL_ECHSTORE *OSSL_ECHSTORE_find_echconfigs(BIO *in);
+```
+
+If the input contains more than one (syntactically correct) ECHConfigList, then only
+those that contain locally supported options (e.g. AEAD ciphers) will be
+returned. If no ECHConfigList found has supported options then none will be
+returned and the function will return NULL.
+
+Additional Client Controls
+--------------------------
+
+Clients can additionally more directly control the values to be used for inner
+and outer SNI and ALPN values via specific APIs. This allows a client to
+override the `public_name` present in an ECHConfigList that will otherwise
+be used for the outer SNI. The `no_outer` input allows a client to emit an
+outer CH with no SNI at all.
+
+```c
+int SSL_ech_set1_server_names(SSL *s, const char *inner_name,
+                             const char *outer_name, int no_outer);
+int SSL_ech_set1_outer_server_name(SSL *s, const char *outer_name, int no_outer);
```

Comment:
Could no_outer be replaced by simply setting outer_name to NULL?

Reply:
'Fraid not. We need to distinguish "use the ECHConfig.public_name" vs. "use this value" vs. "don't send any SNI in outer CH" - I picked NULL for the first (default) case. (Added another sentence to make that clearer.)

Reply:
OK